clash-lib/src/Clash/Backend/VHDL.hs

```diff
@@ -1419,6 +1420,9 @@ inst_ (InstDecl entOrComp libM _ nm lbl gens pms0) = do
 inst_ (BlackBoxD _ libs imps inc bs bbCtx) =
   fmap Just (Ap (column (renderBlackBox libs imps inc bs bbCtx)))
 
+inst_ (ConditionalDecl cond ds) = Just <$>
+  "#ifdef" <+> pretty cond <> line <> insts ds <> line <> "#endif"
+
```

Comment:
For my own understanding: does VHDL support conditional compilation like this? Or is this SymbiYosys only and supported there?

Reply:
Hmm, not sure actually. For symbiyosys it's up to `ghdl`. With this style though any C preprocessor could be used 

Maybe it's best to use 

```vhdl
if FORMAL generate
   some logic using the optional port
end generate;
```

Although I'm not sure how recent a feature this is.

Reply:
`if ... generate` is supported in VHDL '93 which is what Clash aims to produce, so maybe this is a better option? `FORMAL` is defined as a macro in verilog though, AFAIK VHDL doesn't have macros, so I'm wondering if `ghdl` would be passed anything to set `FORMAL` :thinking: 

I'll see if any of my colleagues with more VHDL experience have any suggestions

Reply:
Not necessarily more experience, but a TheInternetSays `if ... generate` is a good option for VHDL 93.

Silly question, but, wouldn't we have seen a GHDL error if it's not valid?

Reply:
The tests added in this PR don't test against VHDL, so possibly not. Having had a brief skim of `ghdl`'s command line options I'm not sure there is a way to pass options like this anyway. If there's no way to pass `FORMAL` without the C preprocessor I wonder if it's a good idea, it means you can no longer necessarily pass VHDL generated by Clash directly to a vendor tool (without running it through `cpp` or manually removing the relevant sections of code first).

Reply:
TBH symbiyosys with ghdl is pretty clunky. The only example I could find of it doesn't do any conditional compilation: https://github.com/tmeissner/psl_with_ghdl/blob/master/src/psl_next.vhd

Reply:
> The tests added in this PR don't test against VHDL, so possibly not. Having had a brief skim of `ghdl`'s command line options I'm not sure there is a way to pass options like this anyway. If there's no way to pass `FORMAL` without the C preprocessor I wonder if it's a good idea, it means you can no longer necessarily pass VHDL generated by Clash directly to a vendor tool (without running it through `cpp` or manually removing the relevant sections of code first).

yeah, that's a good point, perhaps better to just fail compilation with clash, or not emit any vhdl and emit a warning?

Reply:
I think emitting a warning seems more reasonable than throwing out everything. Other parts of the VHDL backend use tracing functions for warnings, so maybe:

```haskell
inst_ ConditionalDecl{} = do
  traceM "WARNING: Conditional compilation is not supported in VHDL, any conditionally compiled code will be discarded."
  return Nothing
```

Reply:
You could also then add another `runTest` for the tests on VHDL only which expects Clash to "fail" with an exit code of 0, and the warning printed to stderr

Reply:
Done, I mentioned the condition in the message to for greater clarity.

Reply:
> You could also then add another `runTest` for the tests on VHDL only which expects Clash to "fail" with an exit code of 0, and the warning printed to stderr

hmm, this is tricky actually because currently `YosysSVA` already selects PSL in comments for VHDL, and never actually generates any ConditionalDecls.

Reply:
(and there's no other way of generating ConditionalDecls for vhdl)

Reply:
Ah, that's fine then. I assumed something in the new tests would already hit the warning